### PR TITLE
Geo Lookup Updates

### PR DESCRIPTION
**Pull request by AntoxaAntoxic**
https://github.com/prebid/prebid-server/issues/3392
https://github.com/prebid/prebid-server/issues/3559

### CODE REVIEW

File: src/main/java/org/prebid/server/auction/privacy/contextfactory/AmpPrivacyContextFactory.java [hunk 1]
```diff
@@ -54,14 +54,16 @@ public Future<PrivacyContext> contextFrom(AuctionContext auctionContext) {
         final Privacy strippedPrivacy = stripPrivacy(initialPrivacy, auctionContext);
         final Device device = bidRequest.getDevice();
 
+        //todo: pass geo info here as well
```
Comment (AntoxaAntoxic) [inline src/main/java/org/prebid/server/auction/privacy/contextfactory/AmpPrivacyContextFactory.java:57]: remove

File: src/main/java/org/prebid/server/deals/UserAdditionalInfoService.java [hunk 2]
```diff
@@ -101,6 +101,7 @@ private Future<DeviceInfo> lookupDeviceInfo(Device device) {
                 : Future.failedFuture("Device info is disabled by configuration");
     }
 
+    //todo: to be removed
```
Comment (AntoxaAntoxic) [inline src/main/java/org/prebid/server/deals/UserAdditionalInfoService.java:104]: remove

File: src/main/java/org/prebid/server/auction/requestfactory/AmpRequestFactory.java [hunk 3]
```diff
@@ -142,6 +146,12 @@ public Future<AuctionContext> fromRequest(RoutingContext routingContext, long st
 
                 .map(auctionContext -> auctionContext.with(debugResolver.debugContextFrom(auctionContext)))
 
+                .compose(auctionContext -> geoLocationServiceWrapper.lookup(auctionContext)
+                        .map(auctionContext::with))
+
+                .map(auctionContext -> auctionContext.with(
+                        ortb2RequestFactory.enrichBidRequestWithGeolocationData(auctionContext)))
```
[the remark this answers is not in the document]
Comment (AntoxaAntoxic) [reply src/main/java/org/prebid/server/auction/requestfactory/AmpRequestFactory.java:153]: imho it's in the same style, I need first to populate geoInfo and then enrich the device, the geoInfo is needed later for the TCF (and for other future features), thus there two steps here

File: src/main/java/org/prebid/server/privacy/gdpr/TcfDefinerService.java [hunk 4]
```diff
@@ -205,17 +216,9 @@ private Future<TcfContext> prepareTcfContext(Privacy privacy,
...
-        if (ipAddress != null && geoLocationService != null) {
-            return geoLocationService.lookup(effectiveIpAddress, timeout)
-                    .map(geoInfo -> updateMetricsAndEnrichWithGeo(geoInfo, defaultContext))
-                    .recover(error -> logError(error, defaultContext));
-        }
-
-        return Future.succeededFuture(defaultContext);
+        return geoLocationServiceWrapper.doLookup(effectiveIpAddress, country, timeout)
+                .recover(ignored -> Future.succeededFuture(geoInfo))
+                .map(lookupResult -> enrichWithGeoInfo(defaultContext, lookupResult, country));
```
[the remark this answers is not in the document]
Comment (AntoxaAntoxic) [reply src/main/java/org/prebid/server/privacy/gdpr/TcfDefinerService.java:221]: I don't do the lookup the second time, the wrapper will decide whether it's necessary to look up based on the presence of country, ip etc.

File: src/main/java/org/prebid/server/privacy/gdpr/TcfDefinerService.java [hunk 5]
```diff
@@ -205,17 +216,9 @@ private Future<TcfContext> prepareTcfContext(Privacy privacy,
             return Future.succeededFuture(defaultContext.toBuilder().inGdprScope(inScopeOfGdpr(gdpr)).build());
         }
 
-        if (country != null) {
-            return Future.succeededFuture(defaultContext.toBuilder().inGdprScope(inScopeOfGdpr(inEea)).build());
-        }
```
[the remark this answers is not in the document]
Comment (AntoxaAntoxic) [reply src/main/java/org/prebid/server/privacy/gdpr/TcfDefinerService.java:210]: it happens in the `enrichWithGeoInfo`

File: src/main/java/org/prebid/server/spring/config/GeoLocationConfiguration.java [hunk 6]
```diff
@@ -198,4 +200,17 @@ private String readCsv(Resource resource) throws IOException {
...
+    @Configuration
+    static class GeoLocationWrapperConfiguration {
+
+        @Bean
+        GeoLocationServiceWrapper geoLocationServiceWrapper(
+                @Autowired(required = false) GeoLocationService geoLocationService,
+                Metrics metrics) {
+
+            return new GeoLocationServiceWrapper(geoLocationService, metrics);
+        }
```
[the remark this answers is not in the document]
Comment (AntoxaAntoxic) [reply src/main/java/org/prebid/server/spring/config/GeoLocationConfiguration.java:213]: the parent class is not a configuration annotated, thus I thought the style here is making own configuration

File: src/main/java/org/prebid/server/auction/requestfactory/AmpRequestFactory.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (CTMBNara) [reply src/main/java/org/prebid/server/auction/requestfactory/AmpRequestFactory.java:153]: It's not about combining two actions... :) I just suggested refactoring like this:
```
                .compose(auctionContext -> ortb2RequestFactory.enrichBidRequestWithGeolocationData(auctionContext)
                        .map(auctionContext::with))
```

File: src/main/java/org/prebid/server/privacy/gdpr/TcfDefinerService.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (CTMBNara) [reply src/main/java/org/prebid/server/privacy/gdpr/TcfDefinerService.java:221]: But you do it a second time if the first lookup fails...

File: src/main/java/org/prebid/server/spring/config/GeoLocationConfiguration.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (CTMBNara) [reply src/main/java/org/prebid/server/spring/config/GeoLocationConfiguration.java:213]: Later will you spam child configuration classes for each bean?

File: src/main/java/org/prebid/server/privacy/gdpr/TcfDefinerService.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (AntoxaAntoxic) [reply src/main/java/org/prebid/server/privacy/gdpr/TcfDefinerService.java:221]: is it a bad thing?

anyway there is another case, when we have a country in the request already, in this case we won't do a geo lookup at all, so we checking the presence of the country is preferable

File: src/main/java/org/prebid/server/spring/config/GeoLocationConfiguration.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (AntoxaAntoxic) [reply src/main/java/org/prebid/server/spring/config/GeoLocationConfiguration.java:213]: I mean, I don't think it's good, I'll fix, no problem)

File: src/main/java/org/prebid/server/privacy/gdpr/TcfDefinerService.java [hunk 7]
```diff
@@ -237,28 +240,18 @@ private static boolean shouldMaskIp(TCString consent) {
         return isConsentValid(consent) && consent.getVersion() == 2 && !consent.getSpecialFeatureOptIns().contains(1);
     }
 
-    private TcfContext updateMetricsAndEnrichWithGeo(GeoInfo geoInfo, TcfContext tcfContext) {
-        metrics.updateGeoLocationMetric(true);
-        final Boolean inEea = isCountryInEea(geoInfo.getCountry());
+    private TcfContext enrichWithGeoInfo(TcfContext defaultTcfContext, GeoInfo geoInfo, String defaultCountry) {
+        final String country = ObjectUtil.getIfNotNullOrDefault(geoInfo, GeoInfo::getCountry, () -> defaultCountry);
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/privacy/gdpr/TcfDefinerService.java:244]:
```
Optional.ofNullable(geoInfo)
      .map(GeoInfo::getCountry)
      .orElse(defaultCountry)
```

**Review by CTMBNara — APPROVED** [review]

**Review by CTMBNara — APPROVED** [review]

**Pull request merged**